documentation/specs/rar-as-service.md

```diff
@@ -0,0 +1,217 @@
+# Resolve Assembly Reference as Service Design
+
+This document describes the Resolve Assembly Reference as Service
+
+# Background
+
+[MSBuild](https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild?view=vs-2019) is a universal build engine, used for building pretty much everything in the Microsoft world. It is available on command line (msbuild, [dotnet build](https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-build)), runs under the covers when building projects and solutions in Visual Studio, and is used as the local build engine in "higher-order" distributed build systems. Essentially all .NET applications use MSBuild as their primary build engine.
+
+RAR is the acronym behind ResolveAssemblyReference (an MSBuild task) and ResolveAssemblyReferences (an MSBuild target). RAR is used in all .NET builds. Quoting the official documentation, RAR "_Determines all assemblies that depend on the specified assemblies, including second and nth-order dependencies._"
+
+The RAR task has become very complex and slow over the years. It tends to rank high on the list of MSBuild's performance bottlenecks. There is an inherent cost to walking the assembly reference graph and computing the dependency closure, especially in terms of I/O operations. To address this, the task internally maintains caches, both in-memory and on disk. While it alleviates the problem somewhat, it is still a suboptimal solution because
+
+1. The task runs in build nodes which are generally created as separate processes, one per logical CPU, so the in-memory state is duplicated.
+2. Build nodes have limited lifetime and the in-memory state is lost when they die.
+3. No state exists when the task runs for the given project for the first time.
+
+Nathan Mytelka is working on 3. He is planning to ship pre-computed caches as part of VS/SDK so even clean build will be able to skip walking the reference graph for at least system assemblies.
```

Comment:
I'd remove Nathan's name and either just say that it's being worked on separately, or link to an issue/PR.